lib/Scheduling/PresburgerSchedulers.cpp

```diff
@@ -0,0 +1,231 @@
+    // Offsets for variable types.
+    unsigned problemVarOffset = numObj + 1;
+
+    // Map each operation to a variable representing its start time and make
+    // their start time positive.
+    unsigned var = problemVarOffset;
+    for (Operation *op : prob.getOperations()) {
+      opToVar[op] = var;
+      addLowerBound(var, MPInt(0));
+      ++var;
+    }
+  }
+
+  /// Get the number of columns in the constraint system.
+  unsigned getNumCols() const { return LexSimplex::getNumVariables() + 1; }
+
+  using LexSimplex::addEquality;
+  using LexSimplex::addInequality;
+
+  /// Get the index of the operation in the solver.
+  unsigned getOpIndex(Operation *op) const { return opToVar.lookup(op); }
+
+  /// Create a latency constraint representing the given dependence.
+  /// The constraint is represented as:
+  /// dstOpStartTime >= srcOpStartTime + latency.
+  void createLatencyConstraint(MutableArrayRef<MPInt> row,
+                               Problem::Dependence dep) {
+    // Constraint: dst >= src + latency.
+    Operation *src = dep.getSource();
+    Operation *dst = dep.getDestination();
+    unsigned latency = *prob.getLatency(*prob.getLinkedOperatorType(src));
+    row.back() = -latency; // note the negation
+    if (src !=
+        dst) { // note that these coefficients just zero out in self-arcs.
+      row[opToVar[src]] = -1;
+      row[opToVar[dst]] = 1;
+    }
+  }
+
+  /// Create a cyclic latency constraint representing the given dependence and
+  /// the given dependence distance.
+  /// The constraint is represented as:
+  /// dstOpStartTime >= srcOpStartTime + latency + II * distance.
+  void createCyclicLatencyConstraint(MutableArrayRef<MPInt> row,
+                                     Problem::Dependence dep,
+                                     const MPInt &distance) {
+    createLatencyConstraint(row, dep);
+    row[0] = distance;
+  }
+
+  /// Add a constant lower bound on the variable at position `var`, representing
+  /// the constraint: `var >= bound`.
+  void addLowerBound(unsigned var, const MPInt &bound) {
+    SmallVector<MPInt, 8> row(getNumCols());
+    row[var] = 1;
+    row.back() = -bound;
+    addInequality(row);
+  }
+
+  /// Fix the variable at position `var` to a constant, representing the
+  /// constraint: `var == bound`.
+  void addEqBound(unsigned var, const MPInt &bound) {
+    SmallVector<MPInt, 8> row(getNumCols());
+    row[var] = 1;
+    row.back() = -bound;
+    addEquality(row);
+  }
+
+  // Solve the problem, keeping II integer, but allowing the solutions can be
+  // rational. We use a rational lexicographic simplex solver to do this.
+  // To keep II integer, if we obtain a rational II, we fix the II to the
+  // ceiling of the rational II. Since any II greater than the minimum II
+  // is valid, this is a valid solution.
+  MaybeOptimum<SmallVector<Fraction, 8>> solveRationally() {
+    MaybeOptimum<SmallVector<Fraction, 8>> sample =
+        LexSimplex::findRationalLexMin();
```

Comment:
The II is not part of the objective, right? So the solver may theoretically return any value here, because the II is just a normal decision variable if I understand correctly.

I would have expected that you optimize for the smallest II first, ceil/fix it, and then optimize for the start time of the last operation.